Move Color Transform to Viewport

danieldresser-ie opened the pull request:
Requires: https://github.com/ImageEngine/cortex/pull/1284

This seems to be working fairly well.  It is rather lacking in tests however - I did get a Python binding working for the new OpenColorIOAlgo function using pybind, but we don't currently really have any examples of UI tests that check final pixel values after a GL shader has run.

The other thing we may need to think about is what our minimum GL requirements are, and whether the renderbuffer settings I'm requesting are supported on that minimum spec.

There are a couple of misc fixes in here as well - most importantly, "UVView : Pass context to ImageGadgets used for textures" - without that, texture preview in the UV Inspector stopped working in Gaffer 1.0 ( because the default viewName wasn't getting through ).

johnhaddon commented on src/GafferUI/OpenColorIOAlgo.cpp:
```diff
@@ -0,0 +1,359 @@
+	static const char *g_vertexSource =
+	"void main()"
+	"{"
+	"   gl_Position = gl_Vertex;"
+	"   gl_TexCoord[0] = gl_MultiTexCoord0;"
```
Can we avoid the legacy pipeline here please? GafferSceneUI/OutputBuffer.cpp shows that it's pretty trivial for this simple use case.

johnhaddon commented on include/GafferSceneUI/UVView.h:
```diff
@@ -127,6 +127,10 @@ class GAFFERSCENEUI_API UVView : public GafferUI::View
 
 		bool m_framed;
 
+		using DisplayTransformMap = std::unordered_map<std::string, IECoreGL::Shader::SetupPtr>;
```
Do we have any concerns about the GPU memory usage of caching currently-unused transforms and their textures?

danieldresser-ie replied:
I have switched this to caching the ImageProcessor, and regenerating the shader every time ( matching more closely to what was previously happening in TextureGadget ).  Thinking about this though, I should probably flag that currently this is only supporting GPU transforms.  If we decide to move the CPU transform to ImageView, there will be no way for UVView to use the CPU path, since it doesn't use ImageView.

johnhaddon commented on include/GafferImageUI/ImageView.h:
```diff
@@ -147,6 +147,8 @@ class GAFFERIMAGEUI_API ImageView : public GafferUI::View
 				Gaffer::PlugPtr createCounterpart( const std::string &name, Plug::Direction direction ) const override;
 		};
 
+		void setSoloChannel( int soloChannel );
```
Views are intended to define all settings via plugs, _not_ via accessors and member data. The original design was such that ChannelChooser defined a `soloChannel` plug - this plug needs to remain as the only public interface (I presume this implies some refactoring).

danieldresser-ie replied:
I'm not sure I understand enough of the concern to understand what refactoring is required.  I've switched the accessor from public to protected.  It makes sense that it's not part of the public interface, and is only to be called from ChannelChooser.  Is further refactoring required?

johnhaddon replied:
The basic design of the View classes is that there's a central class which does the basics (ImageView in this case), and then "bolt on" classes that add in features (ChannelChooser in this case). Eventually I was thinking this might be exposed so that people could build more customised views (not everything makes sense to have on in the `gaffer view` app vs the `gaffer gui` app for instance). But in the meantime they're still a useful way of structuring the code, hopefully preventing it becoming a complete spaghetti. So I guess the question is whether responsibilities can remain separated, and if not, does ChannelChooser still make sense as a bolt-on component...

johnhaddon commented on include/GafferImageUI/ImageGadget.h:
```diff
@@ -137,11 +137,11 @@ class GAFFERIMAGEUI_API ImageGadget : public GafferUI::Gadget
-
-		void setUseGPU( bool useGPU );
-		bool getUseGPU() const;
+		// Use for setting a display color transform that is a Gaffer image processing node that
+		// will run on the CPU.  This approach is deprecated, it is better to set a GPU OCIO
+		// transform on the ViewportGadget that contains the ImageGadget.
+		void setCPUDisplayTransform( GafferImage::ImageProcessorPtr displayTransform );
+		GafferImage::ConstImageProcessorPtr getCPUDisplayTransform() const;
```
Can we just rip this out now, along with the internal grade nodes? Hosting nodes inside the Gadget was always [heinous](https://github.com/GafferHQ/gaffer/pull/4767/commits/40bdb950ccbebaf171112a8f5b33791bf630e31e#diff-6008d16dbf200b7352302918f1b983ec7e15c5d3bd3734bdd226d8fb5d04ebdaR918-R928), and now we don't need to translate them to the GPU in the ImageGadget, it seems like we have the opportunity to clean all that up and host them inside the ImageView.

johnhaddon replied:
Summary of separate discussion : we will do this in a followup PR.

johnhaddon reviewed (changes requested):
Thanks Daniel! Not the most detailed review from me, but hopefully I've spotted a few worthwhile things. Not sure what's going wrong on Mac, but it's encouraging that it is rendering, so I'm hopeful it's a simple bug rather than a missing GL capability.

The immediate priority for me would be to get the two fix commits I've highlighted into a separate PR so we can release them in `1.0.1.1`...

johnhaddon commented on src/GafferUI/OpenColorIOAlgo.cpp:
```diff
@@ -0,0 +1,359 @@
+	"	{"
+	"		return inPixel;\n"
+	"	}"
+	"	else if( inPixel.a == 0 || !unpremultiply || soloChannel == 3 )"
```
```IECore.Exception: ERROR: 0:71: '==' does not operate on 'float' and 'int'```

To get this running on Mac I needed to replace `0` with `0.0`.

johnhaddon commented on include/GafferSceneUI/UVView.h:
```diff
@@ -127,6 +127,10 @@ class GAFFERSCENEUI_API UVView : public GafferUI::View
 
 		bool m_framed;
 
+		using DisplayTransformMap = std::unordered_map<std::string, IECoreGL::Shader::SetupPtr>;
```
To get this compiling on Mac, I needed to add `#include <unordered_map>`.

johnhaddon reviewed (commented):
Sorry, forgot to mention these small tweaks I needed to get a Mac build going...

danieldresser-ie commented:
Lots of sawdust here at the moment, I just barely got things working with avoiding legacy GL.  Not ready for rereview yet, but I wanted to get something up in case you had time to check the Mac thing.  The only change I've made that I think could make a difference is an explicit call to glViewport ... I don't think this is helpful, but I'm not sure what else could be causing the issue.

Also, question with avoiding legacy GL:  in other places, we shove the vertex attribute buffer into the shader setup, but here we take a const Shader::Setup ( and are going to have an API that returns it again ), which implies we shouldn't be modifying the setup that gets passed in?  For this reason, I've just bound the buffers manually, does that seem reasonable?  ( before you comment on the specifics, I'm aware that I should be scoping the bindings on these buffers differently, got a bit sloppy to get a WIP uploaded at end of day ).

johnhaddon commented:
> I wanted to get something up in case you had time to check the Mac thing. The only change I've made that I think could make a difference is an explicit call to glViewport ... I don't think this is helpful, but I'm not sure what else could be causing the issue.

Getting closer I think! In this recording I first hit `F` to frame, and then do some zooming and panning. The image seems to be in sync with the text and the background this time which is good. But everything is squashed into the lower left of the viewport. I suspect this is going to be a retina display issue of some sort :

https://user-images.githubusercontent.com/1133871/181192710-5dfb5856-f328-4fb8-902f-aab7d177cdb1.mov

> in other places, we shove the vertex attribute buffer into the shader setup, but here we take a const Shader::Setup ( and are going to have an API that returns it again ), which implies we shouldn't be modifying the setup that gets passed in?

Maybe it would be reasonable for `setDisplayTransformOrPostProcessShaderDependingOnYourPointOfView()` to take a non-const `Shader::Setup`, since it is becoming the owner? Alternatively, `Shader::Setup` could provide a copy constructor that could be used internally. It seems a pity to have an abstraction like `Setup` and then to need to go around the back of it...

proberts-cinesite commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -419,6 +427,13 @@ class GAFFERUI_API ViewportGadget : public Gadget
 		UnarySignal m_preRenderSignal;
 		UnarySignal m_renderRequestSignal;
 
+		mutable GLuint m_fbo;
+		mutable Imath::V2i m_framebufferSize;
+		mutable GLuint m_framebufferTexture;
+		mutable GLuint m_rbo;
```
you will need to ensure the opengl context is current when you make the gl calls to dispose of them.

danieldresser-ie commented:
OK, I think I've addressed all specific feedback.  In theory, this might be about good to merge after adding Changes entries, if it's working now on Mac ( with a followup PR to remove the newly added setCPUDisplayTransform and replace it with CPU transform support on ImageView ).

I did add one extra commit for you to take a look at:  "ImageGadget : Switch tile shader to be global instead of per-instance" - during further testing, I thought I might have seen a noticeable slowdown when using a UVView with UDIMs that loads many ImageGadgets at once, and I thought this might have reduced it, which would make it worthwhile ( the code is a bit simpler to me anyway ).  The difference is between a 2 second hang and a 1 second hang, and I haven't got an accurate way to measure though, so I could be imagining things, happy to drop this commit if you prefer.

johnhaddon commented:
I had to make [this small change]( https://github.com/johnhaddon/gaffer/commit/a7278cd1cf95261b76160d16d9c044bb81a55979) to the GLSL for Mac, but with that done your latest push appears to be working! 

I did notice though that the CropWindowTool isn't looking so good now - it is drawing in the main layer, so getting a colour correction applied, which is making the crop rectangle pretty washed out, and reducing the contrast of the masking overlay. I think we probably just want to shift this drawing to an overlay layer?

danieldresser-ie commented:
Merged your glsl fix, moved CropWindowTool to overlay, and removed ChannelChooser ( as discussed, we can't really meaningfully separate it from ImageView )

johnhaddon commented:
> OK, I think I've addressed all specific feedback. In theory, this might be about good to merge after adding Changes entries, if it's working now on Mac

I've been through and resolved all feedback (one remains about indentation), and confirmed that https://github.com/GafferHQ/gaffer/pull/4767/commits/68e07e63785eafb37058f316f4b78ba0c6e90409 works on Mac (I assume the following commits are all just trying to figure out CI issues).

> question with avoiding legacy GL: in other places, we shove the vertex attribute buffer into the shader setup, but here we take a const Shader::Setup ( and are going to have an API that returns it again ), which implies we shouldn't be modifying the setup that gets passed in? For this reason, I've just bound the buffers manually, does that seem reasonable? 

This would make sense, except we're taking a non-const Setup at present. If we're going to take a non-const one, we might as well just call `addVertexAttribute()` on it and save ourselves some juggling.

> WIP MORE AGAIN

This is looking painful! Might be time to fire up a Docker container with our build image so you can debug more directly?

danieldresser-ie commented:
I've fixed the indentation, made the Shader::Setup const ( in place ).  The final commit feels kinda awkward, there's probably a better approach, but it at least deals with the test failures I was seeing on Friday due to GL init not being called yet.

With that sorted, I'm now actually seeing the sort of error my heatwave-fried brain thought I was seeing on Friday:  the CI machine is failing to get the fbo attachments set up properly.  Haven't confirmed yet, but my assumption is that GL_RGBA16F isn't supported for the frame buffer texture.  Supporting more than 8 bits is pretty required for a color transformed viewport, so I guess I probably just need to catch this and refuse if someone tries to call setPostProcessShader?

johnhaddon commented:
> CI machine is failing to get the fbo attachments set up properly. Haven't confirmed yet, but my assumption is that GL_RGBA16F isn't supported for the frame buffer texture. Supporting more than 8 bits is pretty required for a color transformed viewport, so I guess I probably just need to catch this and refuse if someone tries to call setPostProcessShader?

I'd suggest we fall back to 8bit on the CI machines, so we can at least exercise the the post-process code path, even if the fidelity isn't there.

danieldresser-ie commented:
Added fallback to RGBA8 for CI.  Added a couple of Changelog comments ( though not for every commit, since some of these changes really aren't user facing, maybe I should have commented more? ).

Only thing I haven't done yet is squashed "ImageView : Don't initialize shaders until preRender", since it wasn't clear if you've reviewed it yet.

johnhaddon commented:
Looks like we have a build failure on Windows :

```
Creating library lib\GafferUI.lib and object lib\GafferUI.exp
[2256](https://github.com/GafferHQ/gaffer/runs/7663406453?check_suite_focus=true#step:10:2257)
ViewportGadget.obj : error LNK2019: unresolved external symbol "public: __cdecl IECoreGL::Buffer::ScopedBinding::ScopedBinding(class IECoreGL::Buffer const &,unsigned int)" (??0ScopedBinding@Buffer@IECoreGL@@QEAA@AEBV12@I@Z) referenced in function "private: void __cdecl GafferUI::ViewportGadget::renderInternal(enum GafferUI::Gadget::RenderReason,enum GafferUI::Gadget::Layer)const " (?renderInternal@ViewportGadget@GafferUI@@AEBAXW4RenderReason@Gadget@2@W4Layer@42@@Z)
[2257](https://github.com/GafferHQ/gaffer/runs/7663406453?check_suite_focus=true#step:10:2258)
ViewportGadget.obj : error LNK2019: unresolved external symbol "public: __cdecl IECoreGL::Buffer::ScopedBinding::~ScopedBinding(void)" (??1ScopedBinding@Buffer@IECoreGL@@QEAA@XZ) referenced in function "private: void __cdecl GafferUI::ViewportGadget::renderInternal(enum GafferUI::Gadget::RenderReason,enum GafferUI::Gadget::Layer)const " (?renderInternal@ViewportGadget@GafferUI@@AEBAXW4RenderReason@Gadget@2@W4Layer@42@@Z)
[2258](https://github.com/GafferHQ/gaffer/runs/7663406453?check_suite_focus=true#step:10:2259)
lib\GafferUI.dll : fatal error LNK1120: 2 unresolved externals
```

I suspect this is because we need IECOREGL_API [here](https://github.com/ImageEngine/cortex/blob/main/include/IECoreGL/Buffer.h#L77). @ericmehl, could you test that theory and open a Cortex PR if it's correct please?

johnhaddon commented on Changes.md:
```diff
@@ -1,3 +1,13 @@
+API
+---
+
+ViewportGadget : Added setPostProcessShader.  This allows the main layer to be rendered to a framebuffer, and processed by a shader before being displayed.  Useful for applying color transforms on the GPU after rendering.
```
Missing preceding `-`. Missing backtick formatting around `setPostProcessShader()`.

johnhaddon commented on Changes.md:
```diff
@@ -1,3 +1,13 @@
+---
+
+ViewportGadget : Added setPostProcessShader.  This allows the main layer to be rendered to a framebuffer, and processed by a shader before being displayed.  Useful for applying color transforms on the GPU after rendering.
+- GafferImageUI : Added OpenColorIOAlgo::displayTransformToFramebufferShader.  Converts an OCIO processor to a shader suitable for use with setPostProcessShader
```
Missing backtick formatting. Missing period.

johnhaddon commented on Changes.md:
```diff
@@ -1,3 +1,13 @@
+
+ViewportGadget : Added setPostProcessShader.  This allows the main layer to be rendered to a framebuffer, and processed by a shader before being displayed.  Useful for applying color transforms on the GPU after rendering.
+- GafferImageUI : Added OpenColorIOAlgo::displayTransformToFramebufferShader.  Converts an OCIO processor to a shader suitable for use with setPostProcessShader
+- ImageView : Apply color transform on viewport instead of ImageGadget.  Should not impact user visible behaviour, but paves the way for future work.
```
We avoid the weird programmery imperative mood in the Changes file, since it is user facing.

We should mention the breaking changes to the ImageGadget API.

johnhaddon reviewed (commented):
LGTM, apart from the nitpicking in Changes.md. If you could squash the fixes in and merge that would be great. Since we're not shipping Windows yet, and assume it will be a little while till we have a new Windows dependencies package with the Cortex fix to `Buffer::ScopedBinding`, I'm happy to let Windows CI fail in the interim...

danieldresser-ie commented:
OK, I think I've addressed all the fixes to the Changes file, and squashed the fix to ImageView.

danieldresser-ie commented:
Where are we at with this now?  I think it's good to go, other than the Changes file now being out of date.

johnhaddon commented:
> Where are we at with this now? I think it's good to go, other than the Changes file now being out of date.

Yep, but we need to update the dependencies packages to include the IECoreGL features this needs (1D/3D texture parameters in Shader::Setup) and the IECoreGL linking fix on Windows (that is causing the CI error). I'm working on that now, and hopefully @ericmehl will be able to chip in too.

johnhaddon commented:
I've merged #4791 and rebased this PR so that it picks up the new dependencies packages. Will merge this once CI completes.
